pkg/querier/batch/stream.go

```diff
@@ -159,11 +159,49 @@ func (bs *batchStream) merge(batch *chunk.Batch, size int) {
 			populate(batch, rt)
 			batch.Next()
 		} else {
-			if (rt == chunkenc.ValHistogram || rt == chunkenc.ValFloatHistogram) && lt == chunkenc.ValFloat {
-				// Prefer histograms than floats. Take left side if both have histograms.
-				populate(batch, rt)
-			} else {
+			// The two samples have the same timestamp. They may be the same exact sample or different one
+			// (e.g. we ingested an in-order sample and then OOO sample with same timestamp but different value).
+			// We want query results to be deterministic, so we select one of the two side deterministically.
+			var pickLeftSide bool
+
+			switch {
+			// If the sample types are the same, we pick the highest value.
+			case lt == rt:
+				if lt == chunkenc.ValFloat {
+					_, lValue := bs.curr().At()
+					_, rValue := batch.At()
+					pickLeftSide = lValue > rValue
+				} else if lt == chunkenc.ValHistogram {
+					_, lValue := bs.curr().AtHistogram()
+					_, rValue := batch.AtHistogram()
+
+					pickLeftSide = (*histogram.Histogram)(lValue).Sum > (*histogram.Histogram)(rValue).Sum
```

Comment:
Also I'm wondering how strict we'd want to be. In theory you could have a situation where Count is the same, but the buckets are different, so you end up flapping anyway. Meaning that you'd have to compare every field to be 100% deterministic.

Reply:
Comparing every field would introduce massive overhead here I feel.

Reply:
my 2c is that we shouldn't be 100% deterministic. The issue we saw is probably caused by a relabelling rule which drops the `pod` label on series. In reality, having the same count and sum is very unlikely in those scenarios.

I suspect the client can also send duplicate data, but a misconfigured Prometheus is way more common than a custom remote-write client with a bug.